### PR TITLE
Added CAAS application provisioner worker.

### PR DESCRIPTION
**Pull request by hpidcock**
## Description of change

Adds the CAAS application provisioner worker and application watcher with filtering.

## QA steps

Unit Tests

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: worker/caasapplicationprovisioner/application.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+func (a *appWorker) updateState(app caas.Application, force bool) error {
+	// Fetching the units here is to ensure happens-before consistency
+	// on the deletion of units.
+	observedUnits, err := a.facade.Units(a.name)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	st, err := app.State()
+	if errors.IsNotFound(err) {
```
Comment (ycliuhw) [inline worker/caasapplicationprovisioner/application.go:184]: return nil ?

**Review by ycliuhw — APPROVED** [review]
LGTM, 
just a few questions, thank you!

File: caas/kubernetes/provider/application/application.go [hunk 2]
```diff
@@ -143,34 +142,52 @@ func (a *app) Ensure(config caas.ApplicationConfig) (err error) {
...
+				Name:        a.name,
+				Namespace:   a.namespace,
+				Labels:      a.labels(),
+				Annotations: a.annotations(config),
+			},
+			Spec: corev1.ServiceSpec{
+				Selector: a.labels(),
+				Type:     corev1.ServiceTypeClusterIP,
+			},
+		},
```
Comment (ycliuhw) [inline caas/kubernetes/provider/application/application.go:188]: We probably don't want to revert an already port opened svc to the initial version in Ensure here.
So the logic here should create first, and ignore the already exist error.
I am currently working on service stuff, so we can leave it as it is for now.

File: worker/caasapplicationprovisioner/application.go (hunk 1 above)
Comment (hpidcock) [reply worker/caasapplicationprovisioner/application.go:184]: Returning nil here would prevent the call to force cleanup remaining units.

File: caas/kubernetes/provider/application/application.go (hunk 2 above)
Comment (hpidcock) [reply caas/kubernetes/provider/application/application.go:188]: Yep agreed, I'll leave this to you.

**Pull request merged**